cloudinit/sources/DataSourceAzure.py

```diff
@@ -683,10 +683,16 @@ def check_instance_id(self, sys_cfg):
     def _iid(self, previous=None):
         prev_iid_path = os.path.join(
             self.paths.get_cpath('data'), 'instance-id')
-        iid = dmi.read_dmi_data('system-uuid')
+        # Older kernels than 4.15 will have UPPERCASE product_uuid.
+        # We don't want Azure to react to an UPPER/lower difference as a new
+        # instance id as it rewrites SSH host keys.
+        # LP: #1835584
+        iid = dmi.read_dmi_data('system-uuid').lower()
         if os.path.exists(prev_iid_path):
             previous = util.load_file(prev_iid_path).strip()
-            if is_byte_swapped(previous, iid):
+            if previous.lower() == iid:
+                return previous
```

Comment:
`is_byte_swapped` specifically considers `previous_id == current_id` to mean that they _don't_ match. Is this addressing an identified bug in that behaviour?

Reply:
The False returned from `is_byte_swapped` reports that prev_id and cur_id are actually a match and not "byte swapped" values which means that returning previous or current won't show up as a new instance id.

A False from is_byte_swapped means don't return the previous value, just return current which makes no different if previous_id == current_id.   This special check is because we also need to return previous_id if previous_id and current_id are case-insensitive matches of each other. In that event, we always want to return previous (the original uuid detected) whether or not that was uppercase or lowercase to avoid triggering new-instance-id detection in cloud-init.

Reply:
Aha, right, makes sense to me now; thanks!